File: runtime-elixir/server/test/smithy_validator_test.exs
defmodule SmithyValidatorTest do
  use ExUnit.Case, async: true

  # -------------------------------------------------------------------------
  # validate/2
  # -------------------------------------------------------------------------

  describe "validate/2" do
    test "returns :ok when all required fields present" do
      input = %{"name" => "Alice", "email" => "a@example.com"}
      assert SmithyValidator.validate(input, ["name", "email"]) == :ok
    end

    test "returns :ok when required fields list is empty" do
      assert SmithyValidator.validate(%{"x" => 1}, []) == :ok
    end

    test "returns :ok when input has more keys than required" do
      input = %{"name" => "Alice", "email" => "a@example.com", "age" => 30}
      assert SmithyValidator.validate(input, ["name"]) == :ok
    end

    test "returns error for a single missing field" do
      input = %{"email" => "a@example.com"}
      assert SmithyValidator.validate(input, ["name", "email"]) ==
               {:error, {:missing_required_fields, ["name"]}}
    end

    test "returns error listing all missing fields" do
      input = %{}
      assert {:error, {:missing_required_fields, missing}} =
               SmithyValidator.validate(input, ["name", "email"])

      assert Enum.sort(missing) == ["email", "name"]
    end

    test "returns error when input is completely empty" do
      assert {:error, {:missing_required_fields, ["id"]}} =
               SmithyValidator.validate(%{}, ["id"])
    end

    test "preserves field order in the error list" do
      input = %{}
      fields = ["a", "b", "c"]
      assert {:error, {:missing_required_fields, ^fields}} =
               SmithyValidator.validate(input, fields)
    end

    test "works with atom keys" do
      input = %{name: "Alice"}
      assert SmithyValidator.validate(input, [:name]) == :ok
    end

    test "atom keys not found by string keys (strict key matching)" do
      input = %{name: "Alice"}
      assert {:error, {:missing_required_fields, ["name"]}} =
               SmithyValidator.validate(input, ["name"])
    end

    test "returns :ok when input has nil values for required fields" do
      # nil is a valid map value — the key exists
      input = %{"name" => nil}
      assert SmithyValidator.validate(input, ["name"]) == :ok
    end
  end

  # -------------------------------------------------------------------------
  # format/1
  # -------------------------------------------------------------------------

  describe "format/1" do
    test "formats a single missing field" do
      result = SmithyValidator.format({:missing_required_fields, ["name"]})
      assert result == "Missing required fields: name"
    end

    test "formats multiple missing fields joined by comma" do
      result = SmithyValidator.format({:missing_required_fields, ["name", "email"]})
      assert result == "Missing required fields: name, email"
    end

    test "returns a string" do
      result = SmithyValidator.format({:missing_required_fields, ["x"]})
      assert is_binary(result)
    end

    test "starts with 'Missing required fields:'" do
      result = SmithyValidator.format({:missing_required_fields, ["x", "y"]})
      assert String.starts_with?(result, "Missing required fields:")
    end

    test "contains all field names in the output" do
      fields = ["alpha", "beta", "gamma"]
      result = SmithyValidator.format({:missing_required_fields, fields})

      for field <- fields do
        assert String.contains?(result, field)
      end
    end
  end

  # -------------------------------------------------------------------------
  # Integration: validate then format
  # -------------------------------------------------------------------------

  describe "validate/2 → format/1 integration" do
    test "error from validate can be formatted" do
      input = %{"email" => "a@b.com"}
      {:error, reason} = SmithyValidator.validate(input, ["name", "email"])
      message = SmithyValidator.format(reason)
      assert message =~ "name"
      assert String.starts_with?(message, "Missing required fields:")
    end

    test "round-trip for multiple missing fields" do
      {:error, reason} = SmithyValidator.validate(%{}, ["x", "y", "z"])
      message = SmithyValidator.format(reason)
      assert message == "Missing required fields: x, y, z"
    end
  end
end

File: runtime-erlang/server/test/smithy_validator_test.erl
-module(smithy_validator_test).
-include_lib("eunit/include/eunit.hrl").

%%%===================================================================
%%% Test Suite for smithy_validator
%%%===================================================================

%%--------------------------------------------------------------------
%% validate/2 — all required fields present
%%--------------------------------------------------------------------
validate_all_present_test() ->
    Input = #{<<"name">> => <<"Alice">>, <<"age">> => 30},
    ?assertEqual(ok, smithy_validator:validate(Input, [<<"name">>, <<"age">>])).

validate_empty_required_list_test() ->
    Input = #{<<"name">> => <<"Alice">>},
    ?assertEqual(ok, smithy_validator:validate(Input, [])).

validate_empty_input_no_required_test() ->
    ?assertEqual(ok, smithy_validator:validate(#{}, [])).

%%--------------------------------------------------------------------
%% validate/2 — missing fields
%%--------------------------------------------------------------------
validate_single_missing_test() ->
    Input = #{<<"name">> => <<"Alice">>},
    Result = smithy_validator:validate(Input, [<<"name">>, <<"email">>]),
    ?assertEqual({error, {missing_required_fields, [<<"email">>]}}, Result).

validate_all_missing_test() ->
    Input = #{},
    Result = smithy_validator:validate(Input, [<<"name">>, <<"age">>]),
    ?assertMatch({error, {missing_required_fields, _}}, Result),
    {error, {missing_required_fields, Missing}} = Result,
    ?assertEqual(lists:sort([<<"name">>, <<"age">>]), lists:sort(Missing)).

validate_preserves_order_test() ->
    %% Missing fields are reported in the order they appear in RequiredFields
    Input = #{},
    {error, {missing_required_fields, Missing}} =
        smithy_validator:validate(Input, [<<"a">>, <<"b">>, <<"c">>]),
    ?assertEqual([<<"a">>, <<"b">>, <<"c">>], Missing).

validate_present_fields_not_in_missing_test() ->
    Input = #{<<"a">> => 1, <<"c">> => 3},
    {error, {missing_required_fields, Missing}} =
        smithy_validator:validate(Input, [<<"a">>, <<"b">>, <<"c">>]),
    ?assertEqual([<<"b">>], Missing).

%%--------------------------------------------------------------------
%% format/1 — error formatting
%%--------------------------------------------------------------------
format_single_field_test() ->
    Reason = {missing_required_fields, [<<"name">>]},
    ?assertEqual(<<"Missing required fields: name">>, smithy_validator:format(Reason)).

format_multiple_fields_test() ->
    Reason = {missing_required_fields, [<<"name">>, <<"email">>]},
    Result = smithy_validator:format(Reason),
    ?assertEqual(<<"Missing required fields: name, email">>, Result).

format_three_fields_test() ->
    Reason = {missing_required_fields, [<<"a">>, <<"b">>, <<"c">>]},
    Result = smithy_validator:format(Reason),
    ?assertEqual(<<"Missing required fields: a, b, c">>, Result).

format_returns_binary_test() ->
    Reason = {missing_required_fields, [<<"x">>]},
    Result = smithy_validator:format(Reason),
    ?assert(is_binary(Result)).

%%--------------------------------------------------------------------
%% Integration: validate then format
%%--------------------------------------------------------------------
validate_and_format_test() ->
    Input = #{<<"email">> => <<"user@example.com">>},
    {error, Reason} = smithy_validator:validate(Input, [<<"name">>, <<"email">>]),
    Msg = smithy_validator:format(Reason),
    ?assertEqual(<<"Missing required fields: name">>, Msg).
